test/integration/apiserver/unstructured_test.go

```diff
@@ -0,0 +1,95 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"context"
+	"testing"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	appsv1applyconfigurations "k8s.io/client-go/applyconfigurations/apps/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	kubeapiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+// TestRequestObjectConvertibleToUnstructured tests that write requests fail if the request object
+// is not convertible to unstructured. The ability to convert an object to unstructured ensures that
+// it can be encoded as JSON and that field managers can be determined.
+func TestRequestObjectConvertibleToUnstructured(t *testing.T) {
+	for _, raw := range []string{
+		``,
+		`"`,
+		`{`,
+		`[`,
+		`1z`,
+		`z`,
+	} {
+		server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{}, framework.SharedEtcd())
+		defer server.TearDownFn()
+
+		protoConfig := server.ClientConfig
+		protoConfig.ContentConfig.ContentType = runtime.ContentTypeProtobuf
+		protoConfig.ContentConfig.AcceptContentTypes = runtime.ContentTypeProtobuf
+		protoClient, err := clientset.NewForConfig(protoConfig)
+		if err != nil {
+			t.Fatalf("unexpected error creating proto client: %v", err)
+		}
+
+		if _, err := protoClient.AppsV1().ControllerRevisions("default").Create(context.TODO(), &appsv1.ControllerRevision{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "test-revision-create",
+			},
+			Data: runtime.RawExtension{Raw: []byte(raw)},
+		}, metav1.CreateOptions{}); err != nil {
+			t.Logf("create returned non-nil error as expected with rawextension %s: %v", raw, err)
+		} else {
+			t.Error("create unexpectedly returned nil error")
+		}
+
+		if _, err := protoClient.AppsV1().ControllerRevisions("default").Apply(context.TODO(), appsv1applyconfigurations.ControllerRevision("test-revision-apply", "default").
+			WithData(runtime.RawExtension{Raw: []byte(raw)}),
+			metav1.ApplyOptions{}); err != nil {
+			// In this case the error is currently client-side, since apply request
+			// bodies must be encoded as JSON.
+			t.Logf("apply returned non-nil error as expected with rawextension %s: %v", raw, err)
+		} else {
+			t.Error("apply unexpectedly returned nil error")
+		}
+
+		existing, err := protoClient.AppsV1().ControllerRevisions("default").Create(context.TODO(), &appsv1.ControllerRevision{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "test-revision-update",
+			},
+			Data: runtime.RawExtension{Raw: []byte(`{}`)},
```

Comment:
why the succeeding create? Why not test it too whether it fails with `raw`? Why do we need the update?

Reply:
About style: I find it helpful to follow the pattern math proofs use:
```
t.Log("Hypothesis")
....
Show that it behaves as expected.
```
It's much clearer to understand the modality of the steps being done.